api/src/pcapi/routes/backoffice_v3/providers/contexts/allocine.py

```diff
@@ -0,0 +1,71 @@
+import typing
+
+import sqlalchemy as sa
+from werkzeug.exceptions import NotFound
+
+from pcapi.core.offerers import models as offerers_models
+from pcapi.core.providers import models as providers_models
+from pcapi.models import db
+from pcapi.utils.clean_accents import clean_accents
+
+from .. import forms
+from .base import ProviderContext
+
+
+class AllocineContext(ProviderContext):
+    @classmethod
+    def provider_class(cls) -> typing.Type:
+        return providers_models.AllocinePivot
+
+    @classmethod
+    def list_providers(cls, query_string: str | None = None) -> list[providers_models.AllocinePivot]:
+        query = providers_models.AllocinePivot.query.options(
+            sa.orm.joinedload(providers_models.AllocinePivot.venue).load_only(
+                offerers_models.Venue.name,
+            )
+        )
+        if query_string:
+            if query_string.isnumeric():
+                query = query.filter(providers_models.AllocinePivot.venueId == int(query_string))
+            else:
+                query = query.join(providers_models.AllocinePivot.venue).filter(
+                    sa.or_(
+                        sa.func.unaccent(offerers_models.Venue.name).ilike(f"%{clean_accents(query_string)}%"),
+                        providers_models.AllocinePivot.theaterId == query_string,
+                    )
+                )
+        return query.all()
+
+    @classmethod
+    def get_form(cls, provider_id: int | None = None) -> forms.EditAllocineForm:
+        if provider_id is None:
+            return forms.EditAllocineForm()
+
+        provider = providers_models.AllocinePivot.query.get_or_404(provider_id)
+        return forms.EditAllocineForm(
+            venue_id=[provider.venueId],
+            theater_id=provider.theaterId,
+            internal_id=provider.internalId,
+        )
+
+    @classmethod
+    def create_provider(cls, form: forms.EditAllocineForm) -> None:
+        provider = providers_models.AllocinePivot(
+            venueId=form.venue_id.data[0],
+            theaterId=form.theater_id.data,
+            internalId=form.internal_id.data,
+        )
+        db.session.add(provider)
```

Comment:
Par curiosité, pourquoi séparer le db.session.add et le db.session.commit ? C'est souvent plus "clair" d'avoir le add juste avant le commit... (donc le mettre dans blueprint.py dans ce cas)

Reply:
En fait je fais un seul `commit()` dans la partie générique, commune aux quatre implémentations.
Ici dans allocine, il n'y a qu'un seul `add()`, mais je l'ai mis ici parce que pour les 3 autres il y a plusieurs (au moins 2) objets à créer, donc 2 `add()` pour un `commit()` générique ensuite.

Reply:
Good to know !